mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,207 @@
+
+if T.TYPE_CHECKING:
+    from typing_extensions import TypedDict
+
+    from ..interpreter import Interpreter
+    from ..programs import ExternalProgram
+    from . import ModuleState
+
+    class ProjectKwargs(TypedDict):
+        sources: T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        include_directories: T.List[T.Union[str, build.IncludeDirs]]
+        link_with: T.List[T.Union[build.SharedLibrary, build.StaticLibrary]]
+        tags: T.List[str]
+        native: bool
+        gomod_tidy: bool
+
+
+class GoBuildModule(ExtensionModule):
+    INFO = ModuleInfo("Go build", "1.6.0", unstable=True)
+    counter: int = 0
+
+    def __init__(self, interpreter: Interpreter) -> None:
+        super().__init__(interpreter)
+        self.go_exe: T.Union[ExternalProgram, build.Executable, None] = None
+        self.methods.update(
+            {
+                "project": self.project,
+            }
+        )
+
+    def detect_tools(self, state: ModuleState, is_native: bool) -> None:
+        self.go_exe = state.find_program("go")
+
+        if not is_native and state.environment.is_cross_build():
+            state.find_program("x86_64-w64-mingw32-gcc")
+            state.find_program("x86_64-w64-mingw32-g++")
+
+    @typed_pos_args(
+        "gobuild.project",
+        str,
+        varargs=(
+            str,
+            mesonlib.File,
+            build.CustomTarget,
+            build.CustomTargetIndex,
+            build.GeneratedList,
+        ),
+    )
+    @typed_kwargs(
+        "gobuild.project",
+        CT_INPUT_KW.evolve(name="sources"),
+        KwargInfo(
+            "include_directories",
+            ContainerTypeInfo(list, (str, build.IncludeDirs)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
+        KwargInfo(
+            "gomod_tidy",
+            bool,
+            default=True,
+        ),
+    )
+    def project(
+        self,
+        state: ModuleState,
+        args: T.Tuple[
+            str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        ],
+        kwargs: ProjectKwargs,
+    ) -> ModuleReturnValue:
+        if self.go_exe is None:
+            self.detect_tools(state, kwargs["native"])
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs["sources"]))
+        )
+        self.counter += 1
+
+        go_sum_target = build.CustomTarget(
+            f"go mod tidy {self.counter}",
```

Comment:
This counter is rarely used. The only real-world use case is the Wails v2 desktop SDK, which compiles and run `demo.exe` the first time to generate header files, and then the second time to compile the bundled assets.

Instead of counters, maybe we should write the f-string as `f"go mod tidy {proj_name:s}"` instead.

Reply:
Update: counter trick removed in favor of `f"go mod tidy {project_name:s}"`.